### ISSUE TITLE
Documentation confusion around pipelining

### ISSUE THREAD

**Issue by jreed-Aces** [StackExchange/StackExchange.Redis#2380]
While reading this document, there's some confusion in how pipelining is used (unless I'm just misreading). https://stackexchange.github.io/StackExchange.Redis/PipelinesMultiplexers.html

In the examples, the documentation calls out two specific ways to use pipelining. Those are async methods and the "fire and forget" calls. 

However, further down the page in the "Multiplexing" section, the text reads:
> regardless of whether the requests use blocking or asynchronous access, the work is all pipelined.
This seems to say that pipelining will be used regardless of which methods, async or sync, are called to retrieve values.

Am I misreading this? If not, can we have some clarifications on when pipelining is/isn't used? If I have several parallel calls to `StringGet()` on different threads, would this utilize pipelining behind the scenes? The "Pipelining" section seems to say no, while the "Multiplexing" section seems to say yes.

Thanks!

**Comment by NickCraver** [comment]
Pipelining is used in all cases, it's a fundamental of the client. The 2 pieces you're comparing here are whether the work _for that thread_ is being pipelined (it's not if you're synchronously waiting on each round-trip call), and if the _overall work of the client_ is pipelines across all threads, which yes: it always is. Does that help?

**Comment by jreed-Aces** [comment]
Thanks for the response @NickCraver, I really appreciate your help.

I'm a little unclear on what you mean by "the overall work of the client". What work would that be, if not the work being requested on the various threads? Are you talking about general connection management work, or something more?

Related to this, is there any way, either in StackExchange.Redis or in Redis itself, to identify when pipelining is/isn't being used? That would be helpful, but I haven't found a way to do identify that currently. That's caused me to be a bit blind when determining if a patter does/doesn't utilize pipelining correctly.

Thank you again for your help.

**Comment by NickCraver** [comment]
All of the above counts as overall work of the client: work by your threads, connection heartbeats, monitoring, etc. We're sending all those commands for a multiplexer in a pipeline over the shared socket to and from the server. Think of it as a big round-trip conveyor belt. Requests go on one belt on the way to the server, responses come back on a belt back from the server, all in the exact same order...that's how we marry them up.

Pipelining isn't some magical concept overall, it's just the term for sending requests on those convey belts without waiting for each one to go there and back, making the most efficient use of the shared connection resource.

The short version is: use async for async things (e.g. everything Redis - it's off-box), the client will handle the rest.

**Comment by jreed-Aces** [comment]
Thanks @NickCraver. One final clarification to make sure I'm understanding.

I'm working on a legacy application that runs several threads in parallel. All threads share a single ConnectionMultiplexer, and call the non-async method `StringGet()`.

Based on my understanding, none of the `StringGet()` calls on separate threads will utilized pipelining, even if two threads call `StringGet()` at the same moment. However, the underlying connection work that is managed by the ConnectionMultiplexer will utilize pipelining for it's monitoring, heartbeat, and other connection-based work.

However, if I update the code to utilize the `StringGetAsync()` methods, then those calls could also benefit from pipelining along with other Get calls or the general connection management calls.

Is that correct, or am I still misunderstanding.

**Comment by NickCraver** [comment]
That isn't quite correct. 16 synchronous calls to the same multiplexer will be pipelined together by the multiplexer, each thread independently just won't get any benefits. With async, you get a ton of thread pool benefits and could pipeline in a single area if you wanted (e.g. issue 30 commands in a code block before awaiting them all).
